sdk/notificationhubs/notification-hubs/src/client/sendNotification.ts

```diff
@@ -1,33 +1,97 @@
 // Copyright (c) Microsoft Corporation.
 // Licensed under the MIT license.
 
+import {
+  DirectSendNotificationOptions,
+  SendNotificationOptions,
+  isDirectSendNotificationOptions,
+  isSendNotificationOptions,
+} from "../models/options.js";
+import { createRequest, parseNotificationSendResponse, sendRequest } from "./internal/_client.js";
+import { BrowserPushChannel } from "../models/installation.js";
 import { Notification } from "../models/notification.js";
 import { NotificationHubsClientContext } from "./index.js";
 import { NotificationHubsMessageResponse } from "../models/notificationDetails.js";
-import { SendOperationOptions } from "../models/options.js";
-import { sendNotificationPayload } from "./internal/_sendNotificationPayload.js";
+import { createMultipartDirectNotification } from "../utils/notificationUtils.js";
+import { tracingClient } from "../utils/tracing.js";
 
 /**
  * Sends push notifications to devices that match the given tags or tag expression.
  * @param context - The Notification Hubs client.
- * @param tags - The tags used to target the device for push notifications in either an array or tag expression.
  * @param notification - The notification to send to the matching devices.
- * @param options - Configuration options for the direct send operation which can contain custom headers
- * which may include APNs specific such as apns-topic or for WNS, X-WNS-TYPE.
+ * @param options - Options for the notification including tags, device handles and whether to enable test send.
  * @returns A NotificationHubResponse with the tracking ID, correlation ID and location.
  */
 export function sendNotification(
   context: NotificationHubsClientContext,
-  tags: string[] | string,
   notification: Notification,
-  options: SendOperationOptions = {}
+  options: DirectSendNotificationOptions | SendNotificationOptions = { enableTestSend: false }
 ): Promise<NotificationHubsMessageResponse> {
-  return sendNotificationPayload(
-    context,
-    notification,
-    "sendNotification",
-    undefined,
-    tags,
-    options
+  return tracingClient.withSpan(
+    `NotificationHubsClientContext-sendNotification`,
+    options,
+    async (updatedOptions) => {
+      const endpoint = context.requestUrl();
+      endpoint.pathname += "/messages/";
+
+      // Check if batch direct send
+      if (isDirectSendNotificationOptions(options) && Array.isArray(options.deviceHandle)) {
+        endpoint.pathname += "$batch";
+      }
+
+      // Check for test send
+      if (isSendNotificationOptions(options) && options.enableTestSend) {
+        endpoint.searchParams.append("test", "true");
+      }
+
+      const headers = await context.createHeaders("sendNotification");
```

Comment:
nit: since `createHttpHeaders` can take a plain object as input, could pass `notification.headers` along to your `createHeaders` helper if you wanted to avoid the loop

Reply:
Moved code to createHeaders method to set the raw headers.